processing/src/test/java/io/druid/segment/IndexMergerTest.java

```diff
@@ -813,6 +817,66 @@ public void testDisjointDimMerge() throws Exception
     Assert.assertArrayEquals(new Object[]{1L}, boatList.get(4).getMetrics());
   }
 
+  @Test
+  public void testMismatchedMetrics() throws IOException
+  {
+    IncrementalIndex index1 = IncrementalIndexTest.createIndex(new AggregatorFactory[]{
+        new LongSumAggregatorFactory("A", "A")
+    });
+    closer.closeLater(index1);
+
+    IncrementalIndex index2 = IncrementalIndexTest.createIndex(new AggregatorFactory[]{
+        new LongSumAggregatorFactory("A", "A"),
+        new LongSumAggregatorFactory("C", "C")
+    });
+    closer.closeLater(index2);
+
+    IncrementalIndex index3 = IncrementalIndexTest.createIndex(new AggregatorFactory[]{
+        new LongSumAggregatorFactory("B", "B")
+    });
+    closer.closeLater(index3);
+
+    IncrementalIndex index4 = IncrementalIndexTest.createIndex(new AggregatorFactory[]{
+        new LongSumAggregatorFactory("C", "C"),
+        new LongSumAggregatorFactory("A", "A"),
+        new LongSumAggregatorFactory("B", "B")
+    });
+    closer.closeLater(index4);
+
+    IncrementalIndex index5 = IncrementalIndexTest.createIndex(new AggregatorFactory[]{
+        new LongSumAggregatorFactory("C", "C"),
+        new LongSumAggregatorFactory("B", "B")
+    });
+    closer.closeLater(index5);
+
+
+    Interval interval = new Interval(0, new DateTime().getMillis());
+    RoaringBitmapFactory factory = new RoaringBitmapFactory();
+    ArrayList<IndexableAdapter> toMerge = Lists.<IndexableAdapter>newArrayList(
+        new IncrementalIndexAdapter(interval, index1, factory),
+        new IncrementalIndexAdapter(interval, index2, factory),
+        new IncrementalIndexAdapter(interval, index3, factory),
+        new IncrementalIndexAdapter(interval, index4, factory),
+        new IncrementalIndexAdapter(interval, index5, factory)
+    );
+
+    final File tmpDirMerged = temporaryFolder.newFolder();
+
+    INDEX_MERGER.merge(
+        toMerge,
+        new AggregatorFactory[] {
+            new LongSumAggregatorFactory("A", "A"),
```

Comment:
@drcrallen just to clarify, my PR did not aim to handle the case where there are metrics in toMerge that are not in the requested AggregatorFactory[]. It appears that the AggregatorFactory can not be retrieved or derived from the IndexAdaptor. If they could, then we could just remove the AggregatorFactory[] from the merge entirely, and let it derive all of them